lib/src/merged_tree.rs

```diff
@@ -1074,41 +980,105 @@ impl Stream for TreeDiffStreamImpl<'_> {
-                    let (key, result) = entry.remove_entry();
+                    let (path, result) = entry.remove_entry();
                     Poll::Ready(Some(match result {
                         Err(err) => TreeDiffEntry {
-                            path: key.path,
+                            path,
                             values: Err(err),
                         },
                         Ok((before, after)) => TreeDiffEntry {
-                            path: key.path,
+                            path,
                             values: Ok((before, after)),
                         },
                     }))
                 }
-                Ok((before, after)) if !before.is_tree() && !after.is_tree() => {
-                    // A diff with no trees involved
-                    let (key, result) = entry.remove_entry();
+                Ok(_) => {
+                    // A diff with a file on at least one side
+                    let (path, result) = entry.remove_entry();
                     Poll::Ready(Some(match result {
                         Err(err) => TreeDiffEntry {
-                            path: key.path,
+                            path,
                             values: Err(err),
                         },
                         Ok((before, after)) => TreeDiffEntry {
-                            path: key.path,
+                            path,
                             values: Ok((before, after)),
                         },
                     }))
                 }
-                _ => {
-                    // The first entry has a tree on at least one side (before or after). We need to
-                    // wait for that future to complete.
-                    assert!(!self.pending_trees.is_empty());
-                    Poll::Pending
-                }
             }
-        } else {
+        } else if self.pending_trees.is_empty() {
             Poll::Ready(None)
+        } else {
+            Poll::Pending
+        }
+    }
+}
+
+/// Adapts a `TreeDiffStream` to emit a added file at a given path after a
+/// removed directory at the same path.
+struct DiffStreamForFileSystem<'a> {
+    inner: TreeDiffStream<'a>,
+    next_item: Option<TreeDiffEntry>,
+    held_file: Option<TreeDiffEntry>,
+}
+
+impl<'a> DiffStreamForFileSystem<'a> {
+    fn new(inner: TreeDiffStream<'a>) -> Self {
+        Self {
+            inner,
+            next_item: None,
+            held_file: None,
+        }
+    }
+}
+
+impl<'a> Stream for DiffStreamForFileSystem<'a> {
+    type Item = TreeDiffEntry;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        loop {
+            if let Some(next) = self.next_item.take() {
+                // If there's a held file "foo" and the next item to emit is not "foo/...", then
+                // we must be done with the "foo/" directory and it's time to emit "foo" as a
+                // removed file.
```

Comment:
I think it's easier to follow if `poll_next()` is attempted here (than temporarily queuing to `self.next_item`.)

```rust
while let Some(next) = match self.next_item.take() {
    Some(next) => Some(next),
    None => ready!(self.inner.as_mut().poll_next(cx)),
} {
    ...
}
Poll::Ready(self.held_file.take())
```